openmdao/components/meta_model.py

```diff
@@ -147,45 +158,186 @@ def solve_nonlinear(self, params, unknowns, resids):
             `VecWrapper` containing residuals. (r)
         """
         # Train first
+        self._train()
+
+        # Now Predict for current inputs
+        inputs = self._params_to_inputs(params)
+
+        for name, shape in self._surrogate_output_names:
+            surrogate = self._unknowns_dict[name].get('surrogate')
+            if surrogate:
+                unknowns[name] = surrogate.predict(inputs)
+            else:
+                raise RuntimeError("Metamodel '%s': No surrogate specified for output '%s'"
+                                   % (self.pathname, name))
+
+    def _params_to_inputs(self, params, out=None):
+        """
+        Converts from a dictionary of parameters to the ndarray input.
+        """
+        if out is None:
+            inputs = np.zeros(self._input_size)
+        else:
+            inputs = out
+
+        idx = 0
+        for name, sz in self._surrogate_param_names:
+            val = params[name]
+            if isinstance(val, list):
+                val = np.array(val)
+            if isinstance(val, np.ndarray):
+                inputs[idx:idx + sz] = val.flat
+                idx += sz
+            else:
+                inputs[idx] = val
+                idx += 1
+        return inputs
+
+    def jacobian(self, params, unknowns, resids):
+        """
+        Returns the Jacobian as a dictionary whose keys are tuples of the form
+         ('unknown', 'param') and whose values are ndarrays.
+
+        Args
+        ----
+        params : `VecWrapper`
+            `VecWrapper` containing parameters. (p)
+
+        unknowns : `VecWrapper`
+            `VecWrapper` containing outputs and states. (u)
+
+        resids : `VecWrapper`
+            `VecWrapper` containing residuals. (r)
+
+        Returns
+        -------
+        dict
+            Dictionary whose keys are tuples of the form ('unknown', 'param')
+            and whose values are ndarrays.
+        """
+
+        # Ensure the metamodel is trained.
+        self._train()
```

Comment:
Model should always be already trained at this point. As a rule, you can't expect the derivs to be valid until you have called solve_nonlinear on the model.